### ISSUE TITLE
HUDTEXT gives 'Argument is out of range.' when used in boot.

### ISSUE THREAD

**Issue by MisterInSayne** [KSP-KOS/KOS#781]
[LOG 03:02:00.880] kOS: At boot.ks on archive, line 9
	HUDTEXT("Loading bootup in T"+ROUND((TIME-START):SECONDS), 1, 4, 40, rgb(1,1,0.5), FALSE).
                                                                                    ^
Called from boot on archive, line 1
// +----------------+
^

[LOG 03:02:00.883] System.ArgumentOutOfRangeException: Argument is out of range.
Parameter name: index
  at System.Collections.Generic.List`1[System.Object].get_Item (Int32 index) [0x00000] in <filename unknown>:0 
  at kOS.Execution.Stack.Pop () [0x00000] in <filename unknown>:0 
  at kOS.Execution.CPU.PopStack () [0x00000] in <filename unknown>:0 
  at kOS.Execution.CPU.PopValue (Boolean barewordOkay) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Compilation.OpcodePop.Execute (ICpu cpu) [0x00000] in <filename unknown>:0 
  at kOS.Execution.CPU.ExecuteInstruction (kOS.Execution.ProgramContext context) [0x00000] in <filename unknown>:0 
  at kOS.Execution.CPU.ContinueExecution () [0x00000] in <filename unknown>:0 
  at kOS.Execution.CPU.Update (Double deltaTime) [0x00000] in <filename unknown>:0 
[LOG 03:02:00.889] Code Fragment
File                 Line:Col IP   opcode operand
----                 ----:--- ---- --------------------- 
archive/boot.ks         9:39  0066 push $time 
archive/boot.ks         9:44  0067 push $start 
archive/boot.ks         9:43  0068 sub 
archive/boot.ks         9:51  0069 push seconds 
archive/boot.ks         9:51  0070 getmember 
archive/boot.ks         9:51  0071 call roundnearest() 
archive/boot.ks         9:31  0072 add 
archive/boot.ks         9:61  0073 push 1 
archive/boot.ks         9:64  0074 push 4 
archive/boot.ks         9:67  0075 push 40 
archive/boot.ks         9:75  0076 push 1 
archive/boot.ks         9:77  0077 push 1 
archive/boot.ks         9:79  0078 push 0.5 
archive/boot.ks         9:79  0079 call rgb() 
archive/boot.ks         9:85  0080 push False 
archive/boot.ks         9:85  0081 call hudtext() 
archive/boot.ks         9:85  0082 pop <<--INSTRUCTION POINTER--
archive/boot.ks        10:7   0083 push 1 
archive/boot.ks        10:7   0084 wait 
archive/boot.ks        11:1   0085 jump -25 
archive/boot.ks        13:4   0086 push $ship 
archive/boot.ks        13:9   0087 push name 
archive/boot.ks        13:9   0088 getmember 
archive/boot.ks        13:16  0089 push AutoProbeBuddy 
archive/boot.ks        13:14  0090 eq 
archive/boot.ks        13:14  0091 br.false 9 
archive/boot.ks        14:7   0092 push $autoprobebuddy 
archive/boot.ks        14:7   0093 push to 
archive/boot.ks        14:25  0094 push 1 
archive/boot.ks        14:25  0095 call copy() 
archive/boot.ks        15:12  0096 push 1 
archive/boot.ks        15:12  0097 call switch() 
archive/boot.ks        16:2   0098 call 25 

[LOG 03:02:00.901] kOS: Stack dump:
001      SubroutineContext: {CameFromInstPtr 16}
000      SubroutineContext: {CameFromInstPtr 11}

[LOG 03:02:00.903] kOS: Popping context 2
[LOG 03:02:00.904] kOS: Removed Context File                 Line:Col IP   opcode operand
[LOG 03:02:00.905] kOS: Deleting 0 pointers and restoring 0 pointers
[LOG 03:02:00.906] kOS: New current context File                 Line:Col IP   opcode operand

**Comment by MisterInSayne** [comment]
I've tried different ways, but it's hudtext it hangs on each time. without it, everything works fine.

**Comment by Dunbaratu** [comment]
The title says "when used in boot" - are you implying that it's only when used in boot and not if run afterward manually from the terminal?  If so maybe the boot scripts are getting started a bit too early before KSP has all its ducks in a row.

**Comment by MisterInSayne** [comment]
Yes, sorry for the hasty post, it was late and wanted to post it as I came across it before I went to bed :)

When I use "run boot." there aren't any errors. it's only when I "revert to launch" or... and this is why the issue probably isn't ksp... when I use "reboot.".

I've tried using wait, etc nothing makes it working.

**Comment by TDW89** [comment]
Had a bit of a play with this.

Looks like the hudtext displays despite the exception.
The error message seems to think that the error is with the doEcho argument.
If doEcho is true then it is also echoed in the terminal (again despite the exeption).

The exception still happens even if the hudtext is in a separate script called by the boot script both when the second script is just contains the hudtext and when it is passed the arguments by the boot script.
It still gives the exception when i put as many steps as I can think of in:
boot.ks:
```
switch to 0.
run exec("run hudtext(altitude,1,4,40,red,true).").
```
exec.ks:
```
declare parameter code.
log "" to exec.tmp.
delete exec.tmp.
log code to exec.tmp.
run exec.tmp.
```
hudtext.ks:
```
declare parameter text,delay,style,size,colour,doEcho.
hudtext(text,delay,style,size,colour,doecho).
```

**Comment by hvacengi** [comment]
This may have been fixed in some of the local/global scope changes that @Dunbaratu has made recently.  I can't replicate it in the bleeding edge version I have compiled.  Can either of you try using a freshly compiled source, or let us know if we need to provide a download to a fresh version to test?

**Comment by krisdestruction** [comment]
Getting the same issue with my script calling another script. Please provide me with a fresh version to test!

**Comment by krisdestruction** [comment]
My code is as follows.
report.ks:
```
clearscreen.

declare parameter count.

hudtext( "Launch sequence initiated", 1, 2, 100, RGB(1,1,1), true ).
wait 1.

until count <= 0 {
	hudtext( count, 1, 2, 100, GREEN, false ).
	wait 1.
	set count to count - 1.
}

hudtext( "Liftoff!", 1, 2, 100, GREEN, false ).
stage.
```

report.ks:
```
clearscreen.

sas on.
when ship:altitude >= 1000 then {
	print ship:velocity:surface:mag.
}

run liftoff(5).
```

I then proceed to run the script by using `run report.`. I get the following error:

![image](https://cloud.githubusercontent.com/assets/1440068/7154450/3a5dd734-e324-11e4-83b7-41c3992cb936.png)

**Comment by krisdestruction** [comment]
I should also mention that a trivial workaround until the issue is fixed is to copy the called script into the main script.

**Comment by hvacengi** [comment]
You can find my latest compiled version here: https://dl.dropboxusercontent.com/u/19195825/kOS.zip

I'll try to run your exact scripts later today to test.  Yours is a little different from the one I tested, since I didn't use a loop.  It does look like this and many other functions were updated since 16.2 to more smoothly handler returning from functions.  It used to be that the function had to push a dummy return value because the compiler forced a pop immediately after calling a function like this.  That's no longer the case, so it should be fixed, but please verify.

**Comment by krisdestruction** [comment]
Will verify tonight, thanks!

**Comment by krisdestruction** [comment]
When I install your build, it seems that the parts are all missing so I can't test it.

![image](https://cloud.githubusercontent.com/assets/1440068/7177221/e616f4ba-e3f2-11e4-9608-c19066985711.png)

Here's a pastebin of the startup log
http://pastebin.com/qyfDJq1w

**Comment by hvacengi** [comment]
Oops.  Sorry, I forgot that KSPAPIExtensions is now a requirement.  I broke my "clean" install the same way when I first compiled this dev version a couple of weeks ago. https://github.com/Swamp-Ig/KSPAPIExtensions/releases

**Comment by krisdestruction** [comment]
Yup, the issue is fixed and the script runs without errors in that build.

**Comment by hvacengi** [comment]
Once the update gets pushed out you should be able to remove the KSPAPIExtensions, unless you need it for another mod.  It will be distributed in the directory with kOS.  Some other mods still expect it to be installed separate mod folder, but the documentation says it should be installed with the mod using it instead.

**Comment by krisdestruction** [comment]
Excellent, well I installed it in the ksproot/plugin folder and that worked for me. It would be nice if the plugin manager is also packaged into the KOS folder as well allowing better folder encapsulation and abstraction.

**Comment by hvacengi** [comment]
The API Extensions portion of this was fixed when we released 0.17.2 (I can't find documentation of such, but it was the first release after the last comment, and it was the release for 1.0 compatibility).  Prior to that we were unable to replicate in the develop branch, and I have not seen the issue since (and I use hudtext extensively in my scripts).  So unless anyone still has the issue, I think I'm going to close it.
